// gpg-wrapper/lib/gpg.js
// Generated by IcedCoffeeScript 1.7.1-a
(function() {
  var E, GPG, colgrep, iced, ispawn, parse, __iced_k, __iced_k_noop, _gpg_cmd, _log;

  iced = require('iced-coffee-script/lib/coffee-script/iced').runtime;
  __iced_k = __iced_k_noop = function() {};

  colgrep = require('./colgrep').colgrep;

  E = require('./err').E;

  parse = require('pgp-utils').userid.parse;

  ispawn = require('iced-spawn');

  _gpg_cmd = "gpg";

  exports.set_gpg_cmd = function(c) {
    return _gpg_cmd = c;
  };

  exports.get_gpg_cmd = function() {
    return _gpg_cmd;
  };

  _log = null;

  exports.set_log = function(l) {
    return _log = l;
  };

  exports.GPG = GPG = (function() {
    function GPG(opts) {
      var c;
      this.CMD = (c = opts != null ? opts.cmd : void 0) != null ? c : _gpg_cmd;
    }

    GPG.prototype.mutate_args = function(args) {};

    GPG.prototype.test = function(cb) {
      var err, out, ___iced_passed_deferral, __iced_deferrals, __iced_k;
      __iced_k = __iced_k_noop;
      ___iced_passed_deferral = iced.findDeferral(arguments);
      (function(_this) {
        return (function(__iced_k) {
          __iced_deferrals = new iced.Deferrals(__iced_k, {
            parent: ___iced_passed_deferral,
            filename: "/home/max/src/keybase/gpg-wrapper/src/gpg.iced",
            funcname: "GPG.test"
          });
          ispawn.run({
            name: _this.CMD,
            args: ["--version"],
            quiet: true
          }, __iced_deferrals.defer({
            assign_fn: (function() {
              return function() {
                err = arguments[0];
                return out = arguments[1];
              };
            })(),
            lineno: 32
          }));
          __iced_deferrals._fulfill();
        });
      })(this)((function(_this) {
        return function() {
          return cb(err, out);
        };
      })(this));
    };

    GPG.prototype.run = function(inargs, cb) {
      var env, err, out, stderr, ___iced_passed_deferral, __iced_deferrals, __iced_k;
      __iced_k = __iced_k_noop;
      ___iced_passed_deferral = iced.findDeferral(arguments);
      stderr = null;
      this.mutate_args(inargs);
      env = process.env;
      delete env.LANGUAGE;
      inargs.name = this.CMD;
      inargs.eklass = E.GpgError;
      inargs.opts = {
        env: env
      };
      if (_log != null) {
        inargs.log = _log;
      }
      if ((inargs.stderr == null) && inargs.quiet) {
        inargs.stderr = stderr = new ispawn.BufferOutStream();
      }
      (function(_this) {
        return (function(__iced_k) {
          __iced_deferrals = new iced.Deferrals(__iced_k, {
            parent: ___iced_passed_deferral,
            filename: "/home/max/src/keybase/gpg-wrapper/src/gpg.iced",
            funcname: "GPG.run"
          });
          ispawn.run(inargs, __iced_deferrals.defer({
            assign_fn: (function() {
              return function() {
                err = arguments[0];
                return out = arguments[1];
              };
            })(),
            lineno: 47
          }));
          __iced_deferrals._fulfill();
        });
      })(this)((function(_this) {
        return function() {
          if ((typeof err !== "undefined" && err !== null) && (stderr != null)) {
            err.stderr = stderr.data();
          }
          return cb(err, out);
        };
      })(this));
    };

    GPG.prototype.command_line = function(inargs) {
      var v;
      this.mutate_args(inargs);
      v = [this.CMD].concat(inargs.args);
      return v.join(" ");
    };

    GPG.prototype.assert_no_collision = function(id, cb) {
      var args, err, n, out, rows, ___iced_passed_deferral, __iced_deferrals, __iced_k;
      __iced_k = __iced_k_noop;
      ___iced_passed_deferral = iced.findDeferral(arguments);
      args = ["-k", "--with-colons", id];
      n = 0;
      (function(_this) {
        return (function(__iced_k) {
          __iced_deferrals = new iced.Deferrals(__iced_k, {
            parent: ___iced_passed_deferral,
            filename: "/home/max/src/keybase/gpg-wrapper/src/gpg.iced",
            funcname: "GPG.assert_no_collision"
          });
          _this.run({
            args: args,
            quiet: true
          }, __iced_deferrals.defer({
            assign_fn: (function() {
              return function() {
                err = arguments[0];
                return out = arguments[1];
              };
            })(),
            lineno: 64
          }));
          __iced_deferrals._fulfill();
        });
      })(this)((function(_this) {
        return function() {
          if (typeof err !== "undefined" && err !== null) {

          } else {
            rows = colgrep({
              patterns: {
                0: /^[sp]ub$/,
                4: new RegExp("^.*" + id + "$", "i")
              },
              buffer: out,
              separator: /:/
            });
            if ((n = rows.length) > 1) {
              err = new E.PgpIdCollisionError("Found two keys for ID=" + short_id);
            }
          }
          return cb(err, n);
        };
      })(this));
    };

    GPG.prototype.assert_exactly_one = function(short_id, cb) {
      var err, n, ___iced_passed_deferral, __iced_deferrals, __iced_k;
      __iced_k = __iced_k_noop;
      ___iced_passed_deferral = iced.findDeferral(arguments);
      (function(_this) {
        return (function(__iced_k) {
          __iced_deferrals = new iced.Deferrals(__iced_k, {
            parent: ___iced_passed_deferral,
            filename: "/home/max/src/keybase/gpg-wrapper/src/gpg.iced",
            funcname: "GPG.assert_exactly_one"
          });
          _this.assert_no_collision(short_id, __iced_deferrals.defer({
            assign_fn: (function() {
              return function() {
                err = arguments[0];
                return n = arguments[1];
              };
            })(),
            lineno: 82
          }));
          __iced_deferrals._fulfill();
        });
      })(this)((function(_this) {
        return function() {
          if (n !== 1) {
            err = new E.NotFoundError("Didn't find a key for " + short_id);
          }
          return cb(err);
        };
      })(this));
    };

    return GPG;

  })();

}).call(this);
