// keybase-proofs/lib/constants.js
// Generated by IcedCoffeeScript 1.7.1-b
(function() {
  var constants, d, k, v, _ref;

  exports.constants = constants = {
    tags: {
      sig: "signature"
    },
    versions: {
      sig: 1
    },
    sig_types: {
      generic_binding: "generic_binding",
      web_service_binding: "web_service_binding",
      track: "track",
      auth: "auth",
      untrack: "untrack",
      revoke: "revoke"
    },
    proof_types: {
      none: 0,
      keybase: 1,
      twitter: 2,
      github: 3,
      generic_web_site: 1000
    },
    expire_in: 60 * 60 * 24 * 365 * 5,
    short_id_bytes: 27,
    v_codes: {
      NONE: 0,
      OK: 1,
      LOCAL: 2,
      FOUND: 3,
      BASE_ERROR: 100,
      HOST_UNREACHABLE: 101,
      PERMISSION_DENIED: 103,
      FAILED_PARSE: 106,
      HTTP_500: 150,
      BASE_HARD_ERROR: 200,
      NOT_FOUND: 201,
      CONTENT_FAILURE: 202,
      BAD_USERNAME: 203,
      BAD_REMOTE_ID: 204,
      HTTP_300: 230,
      HTTP_400: 240,
      HTTP_OTHER: 260,
      DELETED: 301,
      SERVICE_DEAD: 302,
      BAD_SIGNATURE: 303
    },
    user_agent: "keybase.io UserAgent v0.0.1"
  };

  d = {};

  _ref = constants.proof_types;
  for (k in _ref) {
    v = _ref[k];
    d[v] = k;
  }

  exports.proof_type_to_string = d;

}).call(this);
