// keybase-proofs/lib/revoke.js
// Generated by IcedCoffeeScript 1.8.0-c
(function() {
  var Base, Revoke, constants, iced, __iced_k, __iced_k_noop,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  iced = require('iced-runtime');
  __iced_k = __iced_k_noop = function() {};

  Base = require('./base').Base;

  constants = require('./constants').constants;

  exports.Revoke = Revoke = (function(_super) {
    __extends(Revoke, _super);

    function Revoke() {
      return Revoke.__super__.constructor.apply(this, arguments);
    }

    Revoke.prototype._type = function() {
      return constants.sig_types.revoke;
    };

    Revoke.prototype._v_check = function(_arg, cb) {
      var err, json, ___iced_passed_deferral, __iced_deferrals, __iced_k;
      __iced_k = __iced_k_noop;
      ___iced_passed_deferral = iced.findDeferral(arguments);
      json = _arg.json;
      (function(_this) {
        return (function(__iced_k) {
          __iced_deferrals = new iced.Deferrals(__iced_k, {
            parent: ___iced_passed_deferral,
            filename: "/Users/max/src/keybase/proofs/src/revoke.iced",
            funcname: "Revoke._v_check"
          });
          Revoke.__super__._v_check.call(_this, {
            json: json
          }, __iced_deferrals.defer({
            assign_fn: (function() {
              return function() {
                return err = arguments[0];
              };
            })(),
            lineno: 11
          }));
          __iced_deferrals._fulfill();
        });
      })(this)((function(_this) {
        return function() {
          var _ref, _ref1, _ref2, _ref3;
          if (typeof err === "undefined" || err === null) {
            err = !(((_ref = json.body) != null ? (_ref1 = _ref.revoke) != null ? _ref1.sig_id : void 0 : void 0) != null) && !(((_ref2 = json.body) != null ? (_ref3 = _ref2.revoke) != null ? _ref3.sig_ids : void 0 : void 0) != null) ? new Error("Missing revoke.sig_ids in signature") : void 0;
          }
          return cb(err);
        };
      })(this));
    };

    return Revoke;

  })(Base);

}).call(this);
