// keybase-proofs/lib/sibkey.js
// Generated by IcedCoffeeScript 1.8.0-e
(function() {
  var Base, Sibkey, Subkey, SubkeyBase, constants, _ref,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  Base = require('./base').Base;

  constants = require('./constants').constants;

  _ref = require('./subkey'), Subkey = _ref.Subkey, SubkeyBase = _ref.SubkeyBase;

  exports.Sibkey = Sibkey = (function(_super) {
    __extends(Sibkey, _super);

    Sibkey.prototype.get_field = function() {
      return "sibkey";
    };

    Sibkey.prototype.get_new_key_section = function() {
      return this.sibkey;
    };

    Sibkey.prototype.set_new_key_section = function(s) {
      return this.sibkey = s;
    };

    Sibkey.prototype.get_new_km = function() {
      return this.sibkm;
    };

    Sibkey.prototype._type = function() {
      return constants.sig_types.sibkey;
    };

    Sibkey.prototype.need_reverse_sig = function() {
      return true;
    };

    Sibkey.prototype._v_include_pgp_details = function() {
      return true;
    };

    Sibkey.prototype._required_sections = function() {
      return Sibkey.__super__._required_sections.call(this).concat(["sibkey"]);
    };

    Sibkey.prototype._optional_sections = function() {
      return Sibkey.__super__._optional_sections.call(this).concat(["revoke"]);
    };

    function Sibkey(obj) {
      this.sibkey = obj.sibkey;
      this.sibkm = obj.sibkm;
      Sibkey.__super__.constructor.call(this, obj);
    }

    return Sibkey;

  })(SubkeyBase);

}).call(this);
