// keybase-proofs/lib/scrapers/twitter.js
// Generated by IcedCoffeeScript 1.7.1-a
(function() {
  var BaseScraper, BearerToken, Lock, TwitterScraper, bearer_token, constants, decode, iced, make_ids, sncmp, urlmod, v_codes, __iced_k, __iced_k_noop, _bearer_token,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  iced = require('iced-coffee-script/lib/coffee-script/iced').runtime;
  __iced_k = __iced_k_noop = function() {};

  BaseScraper = require('./base').BaseScraper;

  make_ids = require('../base').make_ids;

  constants = require('../constants').constants;

  Lock = require('../util').Lock;

  v_codes = constants.v_codes;

  decode = require('pgp-utils').armor.decode;

  urlmod = require('url');

  sncmp = function(a, b) {
    if ((a == null) || (b == null)) {
      return false;
    } else {
      a = ("" + a).toLowerCase();
      b = ("" + b).toLowerCase();
      return a === b;
    }
  };

  BearerToken = (function() {
    function BearerToken(_arg) {
      this.base = _arg.base;
      this._tok = null;
      this._created = 0;
      this._lock = new Lock();
      this.auth = this.base.auth;
    }

    BearerToken.prototype.get = function(cb) {
      var body, cred, e, err, now, opts, rc, res, tok, ___iced_passed_deferral, __iced_deferrals, __iced_k;
      __iced_k = __iced_k_noop;
      ___iced_passed_deferral = iced.findDeferral(arguments);
      (function(_this) {
        return (function(__iced_k) {
          __iced_deferrals = new iced.Deferrals(__iced_k, {
            parent: ___iced_passed_deferral,
            filename: "/Users/max/src/keybase/proofs/src/scrapers/twitter.iced",
            funcname: "BearerToken.get"
          });
          _this._lock.acquire(__iced_deferrals.defer({
            lineno: 32
          }));
          __iced_deferrals._fulfill();
        });
      })(this)((function(_this) {
        return function() {
          err = null;
          now = Math.floor(Date.now() / 1000);
          (function(__iced_k) {
            if (((res = _this._tok) == null) || (now - _this._created > _this.auth.lifespan)) {
              _this.base.log("+ Request for bearer token");
              cred = (new Buffer([_this.auth.key, _this.auth.secret].join(":"))).toString('base64');
              opts = {
                url: "https://api.twitter.com/oauth2/token",
                headers: {
                  Authorization: "Basic " + cred
                },
                form: {
                  grant_type: "client_credentials"
                },
                method: "POST"
              };
              (function(__iced_k) {
                __iced_deferrals = new iced.Deferrals(__iced_k, {
                  parent: ___iced_passed_deferral,
                  filename: "/Users/max/src/keybase/proofs/src/scrapers/twitter.iced",
                  funcname: "BearerToken.get"
                });
                _this.base._get_url_body(opts, __iced_deferrals.defer({
                  assign_fn: (function() {
                    return function() {
                      err = arguments[0];
                      rc = arguments[1];
                      return body = arguments[2];
                    };
                  })(),
                  lineno: 51
                }));
                __iced_deferrals._fulfill();
              })(function() {
                if (err != null) {
                  _this.base.logl('error', "In getting bearer_token: " + err.message);
                } else if (rc !== v_codes.OK) {
                  _this.base.logl('error', "HTTP error in getting bearer token: " + rc);
                  err = new Error("HTTP error: " + rc);
                } else {
                  try {
                    body = JSON.parse(body);
                  } catch (_error) {
                    e = _error;
                    _this.base.logl('warn', "Could not parse JSON reply: " + e);
                    err = e;
                  }
                }
                if (err != null) {

                } else if ((tok = body.access_token) == null) {
                  _this.base.logl('warn', "No access token found in reply");
                  err = new Error("Twitter error: no access token");
                } else {
                  _this._tok = tok;
                  _this._created = Math.floor(Date.now() / 1000);
                  res = _this._tok;
                }
                return __iced_k(_this.base.log("- Request for bearer token -> " + err));
              });
            } else {
              return __iced_k();
            }
          })(function() {
            _this._lock.release();
            return cb(err, res);
          });
        };
      })(this));
    };

    return BearerToken;

  })();

  _bearer_token = null;

  bearer_token = function(_arg) {
    var base;
    base = _arg.base;
    if (!_bearer_token) {
      _bearer_token = new BearerToken({
        base: base
      });
    }
    return _bearer_token;
  };

  exports.TwitterScraper = TwitterScraper = (function(_super) {
    __extends(TwitterScraper, _super);

    function TwitterScraper(opts) {
      this.auth = opts.auth;
      TwitterScraper.__super__.constructor.call(this, opts);
    }

    TwitterScraper.prototype._check_args = function(args) {
      if (!(args.username != null)) {
        return new Error("Bad args to Twitter proof: no username given");
      } else if (!(args.name != null) || (args.name !== 'twitter')) {
        return new Error("Bad args to Twitter proof: type is " + args.name);
      } else {
        return null;
      }
    };

    TwitterScraper.prototype.hunt2 = function(_arg, cb) {
      var api_url, err, human_url, i, id_str, json, name, out, proof_text_check, rc, remote_id, text, u, username, ___iced_passed_deferral, __iced_deferrals, __iced_k;
      __iced_k = __iced_k_noop;
      ___iced_passed_deferral = iced.findDeferral(arguments);
      username = _arg.username, name = _arg.name, proof_text_check = _arg.proof_text_check;
      out = {};
      rc = v_codes.OK;
      if ((err = this._check_args({
        username: username,
        name: name
      })) != null) {
        return cb(err);
      }
      u = urlmod.format({
        host: "api.twitter.com",
        protocol: "https:",
        pathname: "/1.1/statuses/user_timeline.json",
        query: {
          count: 100,
          screen_name: username
        }
      });
      (function(_this) {
        return (function(__iced_k) {
          __iced_deferrals = new iced.Deferrals(__iced_k, {
            parent: ___iced_passed_deferral,
            filename: "/Users/max/src/keybase/proofs/src/scrapers/twitter.iced",
            funcname: "TwitterScraper.hunt2"
          });
          _this._get_body_api({
            url: u
          }, __iced_deferrals.defer({
            assign_fn: (function() {
              return function() {
                err = arguments[0];
                rc = arguments[1];
                return json = arguments[2];
              };
            })(),
            lineno: 123
          }));
          __iced_deferrals._fulfill();
        });
      })(this)((function(_this) {
        return function() {
          var _i, _len, _ref;
          _this.log("| search index " + u + " -> " + rc);
          if (rc !== v_codes.OK) {

          } else if ((typeof json === "undefined" || json === null) || (json.length === 0)) {
            rc = v_codes.EMPTY_JSON;
          } else {
            for (i = _i = 0, _len = json.length; _i < _len; i = ++_i) {
              _ref = json[i], text = _ref.text, id_str = _ref.id_str;
              if ((_this.find_sig_in_tweet({
                inside: text,
                proof_text_check: proof_text_check
              })) === v_codes.OK) {
                _this.log("| found valid tweet in stream @ " + i);
                rc = v_codes.OK;
                remote_id = id_str;
                api_url = human_url = _this._id_to_url(username, remote_id);
                out = {
                  remote_id: remote_id,
                  api_url: api_url,
                  human_url: human_url
                };
                break;
              }
            }
          }
          out.rc = rc;
          return cb(err, out);
        };
      })(this));
    };

    TwitterScraper.prototype._id_to_url = function(username, status_id) {
      return "https://twitter.com/" + username + "/status/" + status_id;
    };

    TwitterScraper.prototype._check_api_url = function(_arg) {
      var api_url, username;
      api_url = _arg.api_url, username = _arg.username;
      return api_url.indexOf("https://twitter.com/" + username + "/") === 0;
    };

    TwitterScraper.prototype._validate_text_check = function(_arg) {
      var err, msg, proof_text_check, short_id, signature, _ref;
      signature = _arg.signature, proof_text_check = _arg.proof_text_check;
      _ref = decode(signature), err = _ref[0], msg = _ref[1];
      if (err == null) {
        short_id = make_ids(msg.body).short_id;
        if (proof_text_check.indexOf(" " + short_id + " ") < 0) {
          err = new Error("Cannot find " + short_id + " in " + proof_text_check);
        }
      }
      return err;
    };

    TwitterScraper.prototype.find_sig_in_tweet = function(_arg) {
      var html, inside, m, p, proof_text_check, rc, tweet_p, x;
      inside = _arg.inside, tweet_p = _arg.tweet_p, proof_text_check = _arg.proof_text_check;
      if ((tweet_p != null) && (inside == null)) {
        inside = tweet_p.text();
        html = tweet_p.html();
      } else {
        html = null;
      }
      x = /^(@[a-zA-Z0-9_-]+\s+)/;
      this.log("+ Checking tweet '" + inside + " for signature '" + proof_text_check + "'");
      if (html != null) {
        this.log("| html is: " + html);
      }
      while ((m = inside.match(x)) != null) {
        p = m[1];
        inside = inside.slice(p.length);
        this.log("| Stripping off @prefix: " + p);
      }
      rc = inside.indexOf(proof_text_check) === 0 ? v_codes.OK : v_codes.DELETED;
      this.log("- Result -> " + rc);
      return rc;
    };

    TwitterScraper.prototype.check_status = function(_arg, cb) {
      var $, api_url, div, err, html, p, proof_text_check, rc, remote_id, username, ___iced_passed_deferral, __iced_deferrals, __iced_k;
      __iced_k = __iced_k_noop;
      ___iced_passed_deferral = iced.findDeferral(arguments);
      username = _arg.username, api_url = _arg.api_url, proof_text_check = _arg.proof_text_check, remote_id = _arg.remote_id;
      (function(_this) {
        return (function(__iced_k) {
          __iced_deferrals = new iced.Deferrals(__iced_k, {
            parent: ___iced_passed_deferral,
            filename: "/Users/max/src/keybase/proofs/src/scrapers/twitter.iced",
            funcname: "TwitterScraper.check_status"
          });
          _this._get_url_body({
            url: api_url
          }, __iced_deferrals.defer({
            assign_fn: (function() {
              return function() {
                err = arguments[0];
                rc = arguments[1];
                return html = arguments[2];
              };
            })(),
            lineno: 185
          }));
          __iced_deferrals._fulfill();
        });
      })(this)((function(_this) {
        return function() {
          if (rc === v_codes.OK) {
            $ = _this.libs.cheerio.load(html);
            div = $('.permalink-tweet-container .permalink-tweet');
            if (!div.length) {
              rc = v_codes.FAILED_PARSE;
            } else {
              div = div.first();
              rc = !(sncmp(username, div.data('screenName'))) ? v_codes.BAD_USERNAME : ("" + remote_id) !== ("" + div.data('tweetId')) ? v_codes.BAD_REMOTE_ID : ((p = div.find('p.tweet-text')) == null) || !p.length ? v_codes.MISSING : _this.find_sig_in_tweet({
                tweet_p: p.first(),
                proof_text_check: proof_text_check
              });
            }
          }
          return cb(err, rc);
        };
      })(this));
    };

    TwitterScraper.prototype._get_bearer_token = function(cb) {
      var bt, err, rc, tok, ___iced_passed_deferral, __iced_deferrals, __iced_k;
      __iced_k = __iced_k_noop;
      ___iced_passed_deferral = iced.findDeferral(arguments);
      bt = bearer_token({
        base: this
      });
      (function(_this) {
        return (function(__iced_k) {
          __iced_deferrals = new iced.Deferrals(__iced_k, {
            parent: ___iced_passed_deferral,
            filename: "/Users/max/src/keybase/proofs/src/scrapers/twitter.iced",
            funcname: "TwitterScraper._get_bearer_token"
          });
          bt.get(__iced_deferrals.defer({
            assign_fn: (function() {
              return function() {
                err = arguments[0];
                return tok = arguments[1];
              };
            })(),
            lineno: 215
          }));
          __iced_deferrals._fulfill();
        });
      })(this)((function(_this) {
        return function() {
          rc = typeof err !== "undefined" && err !== null ? v_codes.AUTH_FAILED : v_codes.OK;
          return cb(err, rc, tok);
        };
      })(this));
    };

    TwitterScraper.prototype._get_body_api = function(_arg, cb) {
      var args, body, err, rc, tok, url, ___iced_passed_deferral, __iced_deferrals, __iced_k;
      __iced_k = __iced_k_noop;
      ___iced_passed_deferral = iced.findDeferral(arguments);
      url = _arg.url;
      rc = body = err = null;
      (function(_this) {
        return (function(__iced_k) {
          __iced_deferrals = new iced.Deferrals(__iced_k, {
            parent: ___iced_passed_deferral,
            filename: "/Users/max/src/keybase/proofs/src/scrapers/twitter.iced",
            funcname: "TwitterScraper._get_body_api"
          });
          _this._get_bearer_token(__iced_deferrals.defer({
            assign_fn: (function() {
              return function() {
                err = arguments[0];
                rc = arguments[1];
                return tok = arguments[2];
              };
            })(),
            lineno: 224
          }));
          __iced_deferrals._fulfill();
        });
      })(this)((function(_this) {
        return function() {
          (function(__iced_k) {
            if (err == null) {
              _this.log("| HTTP API request for URL '" + url + "'");
              args = {
                url: url,
                headers: {
                  Authorization: "Bearer " + tok,
                  method: "GET"
                },
                json: true
              };
              (function(__iced_k) {
                __iced_deferrals = new iced.Deferrals(__iced_k, {
                  parent: ___iced_passed_deferral,
                  filename: "/Users/max/src/keybase/proofs/src/scrapers/twitter.iced",
                  funcname: "TwitterScraper._get_body_api"
                });
                _this._get_url_body(args, __iced_deferrals.defer({
                  assign_fn: (function() {
                    return function() {
                      err = arguments[0];
                      rc = arguments[1];
                      return body = arguments[2];
                    };
                  })(),
                  lineno: 233
                }));
                __iced_deferrals._fulfill();
              })(__iced_k);
            } else {
              return __iced_k();
            }
          })(function() {
            return cb(err, rc, body);
          });
        };
      })(this));
    };

    return TwitterScraper;

  })(BaseScraper);

}).call(this);
